### PR TITLE
Use `XmlLint` against tool configurations (PHPUnit, PHPCS, Psalm)

### PR DESCRIPTION
**Pull request by internalsystemerror**
|    Q          |   A
|-------------- | ------
| Documentation | no
| Bugfix        | no
| BC Break      | no
| New Feature   | no
| RFC           | no
| QA            | yes

### Description

This PR adds the use of `xmllint` if added via https://github.com/laminas/laminas-continuous-integration-action/pull/37

Note: I don't personally use psalm, and have not checked that the xmllint command works for this schema. However I obtained the schema from the `psalm.xml` of `mezzio/mezzio`, as I did with PHPUnit and PHPCS, both of which work correctly.

### CODE REVIEW

File: src/create-jobs.js
Comment (froschdesign) [inline src/create-jobs.js]: This may fail if a schema version other than 9.3 is used, e.g. 9.5.

Comment (internalsystemerror) [reply src/create-jobs.js]: I considered that, ideally it would just the schema defined in the XML. Feel free to take over this PR if you have some ideas. I considered relying on the xsd installed into vendor. I'm fairly certain PHPUnit would work that way, but I'm not sure about Psalm. Does that come with an XSD installed to vendor?

Comment (internalsystemerror) [reply src/create-jobs.js]: Pushed a change to use the schemas installed to vendor

**Review by weierophinney — CHANGES_REQUESTED** [review]
My primary concern with this is that if these jobs fail... we've still queued jobs for the tools referenced, which means that they still run, but with known invalid schemas.

I think it might make more sense to inline the xmllint into the commands: `xmllint ... && command ...`. This means that we fail early if the config is incorrect for the jobs running those tools.

**Comment by internalsystemerror** [comment]
Yeah I was looking into being able to set a job as dependent on another as I considered that a better approach. Do you have any ideas on that? Otherwise I will just implement the changes you requested.

**Comment by internalsystemerror** [comment]
The problem with inlining the commands is what/how to do that with the PHPUnit matrix? That's why I considered a better solution to be allowing the jobs to be dependent on other jobs.

**Comment by weierophinney** [comment]
@internalsystemerror I think it may make more sense to validate _in this action_; you can see how @boesing managed it for the `composer.json` in #33 . I think there's similar tools for JS?

**Comment by weierophinney** [comment]
> I think it may make more sense to validate _in this action_

On second thought, maybe not - many of these reference a schema file that's installed with the tooling, which happens via Composer.

So, another possibility: do the linting _as a precursor_ to running the associated command?

**Comment by internalsystemerror** [comment]
~~@weierophinney As it stands, the linting will run as a precursor to the associated commands since the jobs will be created with `xmllint ... && command`.~~ EDIT: This was incorrect.

I've fetched the latest changes since this was made, too.

**Comment by internalsystemerror** [comment]
I have modified the commands to run `xmllint ... && command`.

**Comment by boesing** [comment]
Not 100% sure but if a project has both dist and non-dist, this would no create two jobs, right? 

Plenty of unrelated changes there as well. 😅

**Comment by internalsystemerror** [comment]
> Not 100% sure but if a project has both dist and non-dist, this would no create two jobs, right?

Correct, I believe it would.

> Plenty of unrelated changes there as well. 😅

Sorry, I have format on save. I've corrected these but left in the couple of syntax fixes it added.

**Comment by boesing** [comment]
I wonder if it would be possible to avoid having two files created. 
Maybe with using `test` to check if file exists before linting? 

Pseudo-Code:
```
((test -r phpunit.xml && xmllint phpunit.xml ...) || (test -r phpunit.xml.dist && xmllint phpunit.xml.dist)) && vendor/bin/phpunit
```

**Comment by internalsystemerror** [comment]
Nice idea, that would avoid 2 jobs being created if 2 files exist. Thinking about it though, if a second file is submitted (dist xml along with xml), either this is accidental, in which case having the same job twice would be an indication. Or, it's deliberate, in which case you may wish to run these separately? Honestly I don't see this latter one happening though.

**Comment by boesing** [comment]
> Or, it's deliberate, in which case you may wish to run these separately? Honestly I don't see this latter one happening though.

The only thing which came to my mind would be a `.laminas/pre-install.sh` script which applies changes to a config temporarily for whatever case.
In combination with [this action](https://github.com/stefanzweifel/git-auto-commit-action), it could lead to issues when manipulating the `.dist` (as it is committed) so you would probably copy the `*.dist` file to the non-dist version, apply your changes you want to have for CI (for whatever reasons) and then execute the tool. Doing it this way would prevent the github action to detect a change (as usually the non-dist files are added to `.gitignore`).

Fun fact - we probably going to do exactly this for PHPUnit regarding the recent changes for deprecations to exceptions: https://github.com/laminas/laminas-continuous-integration-action/pull/54

So in fact, we are not creating a new file from the `.dist` but modifying whatever file will be used for phpunit (which is non-dist prior dist when it comes to priority).

File: src/create-jobs.js
Comment (boesing) [inline src/create-jobs.js]: not sure if this is a good one.
This will probably create two checks in projects which do have both files.

----

And again, its not about "how likely that will happen". I've provided an [example](https://github.com/laminas/laminas-ci-matrix-action/pull/31#issuecomment-934876500) on why this could happen and thus, this is not that uncommon.

Comment (internalsystemerror) [reply src/create-jobs.js]: The scope of this was just to run `xmllint` against the various xml files.

I've just been cleaning up old PRs and issues and saw this was still outstanding. I'm unsure what change you want me to make here. If there should be a new script which alters how the .xml vs .xml.dist files are handled, I would suggest that should be a new issue with its own PR, and this blocked until that one is merged and rebased after? If that is sufficient, I can create a new PR to do that when I get some time (won't be soon).

File: src/create-jobs.js
Comment (internalsystemerror) [inline src/create-jobs.js]: These commands need to be concatenated with `&&`

File: src/create-jobs.js
Comment (boesing) [inline src/create-jobs.js]: can we rename this to `appendDistExtensionIfFileExists` instead?

**Review by boesing — CHANGES_REQUESTED** [review]
LGTM. Just a lil nitpick but from my perspective, this could work. 
@weierophinney do you see anything else here? I think it would be fine if we spawn the Jobs and let them fail then since we need any feedback from the tooling. 
Just not sure about the fact if we should provide the output from the linter to help devs fixing the problem

File: src/create-jobs.js
Comment (internalsystemerror) [reply src/create-jobs.js]: yep, I'm always bad with names so happy to yield on this.

Comment (boesing) [reply src/create-jobs.js]: same here, but I learned from the best.
/cc @weierophinney 😁

Comment (internalsystemerror) [reply src/create-jobs.js]: My guess is the naming suggestion is to be more descriptive so I went with `appendDistIfFilenameDoesntExist` to clarify that the non-dist version is preferred if both exist.

**Review by boesing — APPROVED** [review]

File: src/create-jobs.js
Comment (weierophinney) [inline src/create-jobs.js]: Is this correct? fs.existsSync is synchronous, so we'd just get back the return value, wouldn't we?

Comment (weierophinney) [reply src/create-jobs.js]: Never mind... just saw we've ALWAYS returned a function that returns boolean. :blush:

File: src/create-jobs.js
Comment (weierophinney) [inline src/create-jobs.js]: Change the name of the constant to "configFileName" or "phpunitConfigFile"; "fileName" is to ambiguous.

File: src/create-jobs.js
Comment (weierophinney) [inline src/create-jobs.js]: Why `--noout`? It seems like we may want to see _why_ it failed.

File: src/create-jobs.js
Comment (weierophinney) [inline src/create-jobs.js]: Same comment as previously: use something like "phpcsConfigFile" or "configFile".

File: src/create-jobs.js
Comment (weierophinney) [inline src/create-jobs.js]: "psalmConfigFile"

**Review by weierophinney — CHANGES_REQUESTED** [review]
Two categories of changes:

- Instead of `const filename = appendDistExtensionIfFilenameDoesntExist(...)`, name the constant something that is semantically appropriate: phpunitConfigFile, phpcsConfigFile, etc.
- I'm thinking we shouldn't use the `--noout` flag. If we have `--noout` and xmllint fails, it's not immediately obvious what happened. With output, the user will know what xmllint errored on.

File: src/create-jobs.js
Comment (internalsystemerror) [reply src/create-jobs.js]: The docs suggest this only suppresses output, not errors, but I will remove anyway.

File: src/create-jobs.js
Comment (internalsystemerror) [reply src/create-jobs.js]: Yeah, there are a couple of fixes I made to the doc types in this manner just because I noticed them.

**Comment by boesing** [comment]
The CI container now supports `before_script` and `after_script`.
We could add the linting as one of the `before_script` entries (its a list of commands).

**Comment by internalsystemerror** [comment]
I've been looking at reimplementing this today, using the new `before_script` option. I was mainly looking for a way to only run the xml linting once per file. For example, using `before_script` on PHPUnit, it would run `xmllint` for each `phpunit` job in the matrix.

Does anyone have any ideas on that?

**Comment by Ocramius** [comment]
Wouldn't it be better to run an independent matrix job that only runs xmllint/yamllint for root level xml/yaml files? 🤔

A completely independent failure would be nicer than mixing it with test results, or worse with no tests run at all.

**Comment by internalsystemerror** [comment]
> A completely independent failure would be nicer than mixing it with test results, or worse with no tests run at all.

Preventing tests from running with a known incorrect schema was requested in this comment https://github.com/laminas/laminas-ci-matrix-action/pull/31#pullrequestreview-662187929.

Adding in these checks as just another linting check would certainly be the easiest solution to implement.

**Comment by weierophinney** [comment]
> I was mainly looking for a way to only run the xml linting once per file.

I think the only way to do that is if we were to do this as one or more separate jobs in the workflow, marked to fail the workflow if it fails. Since we can extend workflows now, we could do this in the main workflow in the .github repo...

**Comment by boesing** [comment]
I am not sure if its smart to have linting for configs seperately. I would expect all Jobs related to that Tool failing if the config is invalid rather than having another Job just to lint. 

So I would be +1 for adding these lint checks as a `before_scripts` entry.

**Comment by Ocramius** [comment]
Works for me: no hard opinion, as long as linting manages to make it through, at least :+1:

**Comment by internalsystemerror** [comment]
OK. Then since I have no preference either, I'll update this PR to simply add xmllint as a `before_script` check for each tool, hopefully this weekend.

**Comment by internalsystemerror** [comment]
Forgive the incoming update spam. This took much longer than I want to admit, mainly because I can't get a reliable way to run the tests even on a clean branch.

Pushing now to get test results via GHA, expect fixes (which is updating the expected `matrix.json` to include the new `before_script` commands) shortly.

**Comment by internalsystemerror** [comment]
Checks are green. Ready for review. @boesing tagging you specifically as you had some ideas already for this.

Matrices that are generated for tools will include `before_script: []` (and the XML lint command where appropriate). Those jobs which are set as explicit/additional checks within `.laminas-ci.json` do not include this feature as I was having issues so decided it was out of scope 😄 .

**Comment by boesing** [comment]
I hope I find some time this weekend to review this.
I wonder if we need some generic stuff like `configurationType: XML|YAML|JSON` so that `xmllint` or `yamllint` does not need to get repeated.

And there is still that "if phpunit.xml exists, we should lint that file, else phpunit.xml.dist". But I would assume not to lint both at the same time for example. But I have no clue how to implement that properly, as the `phpunit.xml` might get created within the container. I did that somewhere to actually dynamically set `convertDeprecationsToExceptions` in PHPUnit Configuration to `false`. Maybe @Ocramius can remember that PR as he reviewed that 😂

**Comment by Ocramius** [comment]
https://www.youtube.com/watch?v=DFdDNOnGjWc

**Comment by internalsystemerror** [comment]
It's been waiting 15 months, another few weeks or even months wouldn't hurt it. The approach I took was to KISS. The tools are modified to reduce the `filesToCheck: string[]` to contain only the files that exist, with empty tools being removed.

Functionally it's the same as before, however the ability to add a tool in the future without any files to check has been removed. Then for each file to check that remains, if there is a linting command available it will add it as a `before_script`.

Any more than that I found myself overengineering something that might not need to exist. For example, I'd also like to extend the before_script functionality to include explicit checks that have their own.

File: src/action/github.ts
Comment (boesing) [inline src/action/github.ts:5]: I am pretty sure that I had to do this.
But since tests are passing, I think its okay now.

Maybe that was on my macos machine when I compiled that locally. Dunno...

**Review by Ocramius — APPROVED** [review]
LGTM, althought to be noted that XMLLint could (in theory) rely on the pre-defined XSD references within `phpunit.xml(.dist)`

**Pull request merged**
